ursula_cli/ursula_inventory.py

```diff
@@ -0,0 +1,193 @@
+#!/usr/bin/env python
+# vim: tabstop=4 shiftwidth=4 softtabstop=4
+
+# Copyright 2016, Jesse Keating <jkeating@j2solutions.net>
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+
+import argparse
+import collections
+import json
+import logging
+import os
+import sys
+import yaml
+from ConfigParser import ConfigParser, NoOptionError, NoSectionError
+
+import ansible
+
+from ansible.constants import DEFAULTS, get_config, load_config_file
+from ansible.inventory import Inventory
+from ansible.inventory.dir import InventoryDirectory, get_file_parser
+from ansible.inventory.group import Group
+from ansible.parsing.dataloader import DataLoader
+from ansible.plugins import vars_loader
+from ansible.utils.vars import combine_vars
+from ansible.utils.vars import load_extra_vars
+from ansible.utils.vars import load_options_vars
+from ansible.vars import VariableManager
+
+LOG = logging.getLogger(__name__)
+
+def deep_update_dict(d, u):
+    for k, v in u.iteritems():
+        if isinstance(v, collections.Mapping):
+            r = deep_update_dict(d.get(k, {}), v)
```

Comment:
This fails when you specify None (e.g. "~") as a default value in `defaults.yml` -- this is something we do in CI.  We factor out as much of our config in `defaults.yml` and for variables that need to be overridden  in group_vars we assign "~" as the default.

```
 def deep_update_dict(d, u):
     for k, v in u.iteritems():
         if isinstance(v, collections.Mapping):
-            r = deep_update_dict(d.get(k, {}), v)
-            d[k] = r
+            dv = d.get(k, {})
+            if dv is None:
+                d[k] = v
+            else:
+                r = deep_update_dict(dv, v)
+                d[k] = r
         else:
             d[k] = u[k]
     return d
```
```